src/rate-limit-throttle.ts

```diff
@@ -0,0 +1,75 @@
+import {isString} from 'lodash'
+import pThrottle from "p-throttle";
+import {AxiosInstance} from "./types";
+
+type ThrottleType = 'auto' | `${number}%`;
+
+const PERCENTAGE_REGEX = /(?<value>\d+)(%)/;
+
+function calculateLimit(type: ThrottleType, max = 7) {
+    let limit = max;
+
+    if (PERCENTAGE_REGEX.test(type)) {
+
+        const groups = type.match(PERCENTAGE_REGEX)?.groups
+        if (groups && groups.value) {
+            const percentage = parseInt(groups.value) / 100;
+            limit = Math.round(max * percentage);
+        }
+    }
+    return Math.min(30, Math.max(1, limit));
+}
+
+function createThrottle(limit: number) {
+    console.log(`Throttle request to ${limit}/s`)
+    return pThrottle({
+        limit,
+        interval: 1000,
+        strict: false
+    });
+}
+
+export default (axiosInstance: AxiosInstance, type: ThrottleType | number = 'auto') => {
+    let limit = isString(type) ? calculateLimit(type) : calculateLimit('auto', type);
+    let throttle = createThrottle(limit);
+    let isCalculated = false;
+
+    let requestInterceptorId = axiosInstance.interceptors.request.use((config) => {
+        return throttle<[], typeof config>(() => config)()
+    }, Promise.reject)
+
+    const responseInterceptorId = axiosInstance.interceptors.response.use((response) => {
+        if (!isCalculated
+            && isString(type) && (type === 'auto' || PERCENTAGE_REGEX.test(type))
```

Comment:
This doesn't have to be calculated on every response.

Reply:
`sCalculated` will be set to `true` after it passes the condition once